### ISSUE TITLE
Decorators are broken in 7

### ISSUE THREAD

**Issue by xixixao** [babel/babel#8459]
## Bug Report

**Current Behavior**
Follow up to https://github.com/babel/babel/issues/7786 . No idea why that issue is closed, with a cliff-hanger. If you add `{legacy: true}` option to the plugin, you'll again get the error about the syntax not being enabled.

**Input Code**
```js
@observer
class TextInput {}
```

**Expected behavior/code**
It works

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  plugins: [
    ...,
    ['@babel/plugin-proposal-decorators', {legacy: true}],
  ],
}
```

**Environment**
- Babel version(s): [v7.0.0-beta.46]
- Node/npm version: [e.g. Node 9.2.1/yarn 1.5.1]
- Monorepo [e.g. no]
- How you are using Babel: [`loader`]

**Additional context/Screenshots**
So basically it goes like this:

First you get a warning `Support for the experimental syntax 'decorators' isn't currently enabled`. You add `'@babel/plugin-proposal-decorators'` to your plugins, you then get a warning about needing `legacy` option, you add the legacy option and you're back at `Support for the experimental syntax 'decorators' isn't currently enabled` warning.

**Comment by nicolo-ribaudo** [comment]
Can you create a reproducible demo?

**Comment by xixixao** [comment]
@nicolo-ribaudo Beyond my knowledge unfortunately.
``` bash
yarn add --dev babel-cli@next
mkdir src
echo "@bla class X {}" > src/test.js
yarn add @babel/plugin-proposal-decorators
# I tried `@babel/plugin-proposal-decorators@next` and yarn let me pick latest beta
# I also tried just the above - this didn't make a difference
./node_modules/.bin/babel --plugins @babel/plugin-proposal-decorators src -d lib
# Output:
./node_modules/.bin/babel --plugins @babel/plugin-proposal-decorators src -d lib
{ Error: [BABEL] src/test.js: Cannot find module '@babel/core'
    at Function.Module._resolveFilename (module.js:542:15)
    at Function.Module._load (module.js:472:25)
    at Module.require (module.js:585:17)
    at require (internal/module.js:11:18)
    at _core (/test-babel/node_modules/@babel/plugin-proposal-decorators/lib/transformer-legacy.js:9:16)
    at Object.<anonymous> (/test-babel/node_modules/@babel/plugin-proposal-decorators/lib/transformer-legacy.js:18:33)
    at Module._compile (module.js:641:30)
    at Object.Module._extensions..js (module.js:652:10)
    at Module.load (module.js:560:32)
    at tryModuleLoad (module.js:503:12) code: 'MODULE_NOT_FOUND' }
```
This is all I could piece together from the existing docs. It would be great to document how to use beta versions.

The bug I'm reporting is happening via `webpack` and `create-react-app` and `react-app-rewired`, but that's not really a minimal demo.

**Comment by nicolo-ribaudo** [comment]
It should be `@babel/cli`, not `babel-cli` (which is an old beta)

**Comment by stevefan1999** [comment]
@xixixao Can confirm.

Chrome stacktrace:
```
Uncaught Error: Module build failed: Thread Loader (Worker 1)
E:\project\presentation\src\index.jsx: Support for the experimental syntax 'decorators' isn't currently enabled (22:3):

> class Store {
>   @observable
>   text = 'Hello World!'
>   @observable
>   counter = 0
> }


Add @babel/plugin-proposal-decorators (https://git.io/vb4ST) to the 'plugins' section of your Babel config to enable transformation.

    at _class.raise (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:779:15)
    at _class.expectOnePlugin (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:2101:18)
    at _class.parseDecorator (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4130:10)
    at _class.parseClassBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4694:30)
    at _class.parseClass (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4656:10)
    at _class.parseStatementContent (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:3991:21)
    at _class.parseStatement (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:3963:17)
    at _class.parseStatement (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:6637:53)
    at _class.parseBlockOrModuleBlockBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4510:23)
    at _class.parseBlockBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4497:10)

    at _class.raise (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:779:15)
    at _class.expectOnePlugin (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:2101:18)
    at _class.parseDecorator (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4130:10)
    at _class.parseClassBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4694:30)
    at _class.parseClass (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4656:10)
    at _class.parseStatementContent (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:3991:21)
    at _class.parseStatement (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:3963:17)
    at _class.parseStatement (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:6637:53)
    at _class.parseBlockOrModuleBlockBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4510:23)
    at _class.parseBlockBody (E:\project\presentation\node_modules\react-scripts\node_modules\@babel\core\node_modules\babylon\lib\index.js:4497:10)
    at Object../src/index.jsx (http://localhost:3000/static/js/main.chunk.js:21:7)
    at __webpack_require__ (http://localhost:3000/static/js/bundle.js:768:30)
    at fn (http://localhost:3000/static/js/bundle.js:142:20)
    at Object../src/index.js (http://localhost:3000/static/js/main.chunk.js:10:1)
    at __webpack_require__ (http://localhost:3000/static/js/bundle.js:768:30)
    at fn (http://localhost:3000/static/js/bundle.js:142:20)
    at Object.0 (http://localhost:3000/static/js/main.chunk.js:34:18)
    at __webpack_require__ (http://localhost:3000/static/js/bundle.js:768:30)
    at checkDeferredModules (http://localhost:3000/static/js/bundle.js:44:23)
    at Array.webpackJsonpCallback [as push] (http://localhost:3000/static/js/bundle.js:31:19)
    at http://localhost:3000/static/js/main.chunk.js:1:57
./src/index.jsx @ index.js:1
__webpack_require__ @ bootstrap:767
fn @ bootstrap:141
./src/index.js @ index.js:1
__webpack_require__ @ bootstrap:767
fn @ bootstrap:141
0 @ index.js:1
__webpack_require__ @ bootstrap:767
checkDeferredModules @ bootstrap:43
webpackJsonpCallback @ bootstrap:30
(anonymous) @ main.chunk.js:1
```

config-overrides.js:
```js
const { injectBabelPlugin, compose } = require('react-app-rewired')
module.exports = {
  webpack (config, env) {
    return compose(
      _ => injectBabelPlugin('react-hot-loader/babel', _),
      _ => injectBabelPlugin('babel-plugin-transform-react-pug', _),
      _ => injectBabelPlugin('babel-plugin-emotion', _),
      _ =>
        injectBabelPlugin(
          ['@babel/plugin-proposal-class-properties', { loose: true }],
          _
        ),
      _ =>
        injectBabelPlugin(
          [
            '@babel/plugin-proposal-decorators',
            {
              legacy: true
            }
          ],
          _
        ),
      _ => {
        _.resolve = Object.assign(_.resolve, {
          symlinks: false,
          mainFields: ['browser', 'main', 'module']
        })
        _.resolve.extensions = _.resolve.extensions.filter(
          _ => _.indexOf('mjs') === -1
        )
        return _
      }
    )(config, env)
  }
}
```

If I remove the `legacy` option, it tells me to put it back on, and if I put it back on, it tells me nope, there's no decorator plugin enabled "apparently", so this Catch-22 really confuses me that I clearly have `@babel/plugin-proposal-decorators` and `legacy` correct but Babylon still told me to toss off...I think I have to ask god what happened

**Comment by nicolo-ribaudo** [comment]
This is probably a configuration issue: probably the plugin isn't applied to some files. Can you share a repository which shows the issue?

**Comment by stevefan1999** [comment]
@nicolo-ribaudo I think not, I tried to dump my webpack config and got this:
```js
module:
   { strictExportPresence: true,
     rules:
      [ { parser: { requireEnsure: false } },
        { test: /\.(js|jsx|mjs)$/,
          enforce: 'pre',
          use:
           [ { options:
                { formatter: [Function: formatter],
                  eslintPath:
                   'E:\\project\\presentation\\node_modules\\eslint\\lib\\api.js',
                  baseConfig:
                   { extends:
                      [ 'E:\\project\\presentation\\node_modules\\react-scripts\\node_modules\\eslint-config-react-app\\index.js' ] },
                  ignore: false,
                  useEslintrc: false },
               loader:
                'E:\\project\\presentation\\node_modules\\eslint-loader\\index.js' } ],
          include: [ 'E:\\project\\presentation\\src' ],
          exclude: [ /[\/\\\\]node_modules[\/\\\\]/ ] },
        { oneOf:
           [ { test: [ /\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/ ],
               loader:
                'E:\\project\\presentation\\node_modules\\url-loader\\dist\\cjs.js',
               options: { limit: 10000, name: 'static/media/[name].[hash:8].[ext]' } },
             { test: /\.(js|jsx|mjs)$/,
               include: [ 'E:\\project\\presentation\\src' ],
               exclude: [ /[\/\\\\]node_modules[\/\\\\]/ ],
               use:
                [ { loader:
                     'E:\\project\\presentation\\node_modules\\thread-loader\\dist\\cjs.js',
                    options: { poolTimeout: Infinity } },
                  { loader:
                     'E:\\project\\presentation\\node_modules\\react-scripts\\node_modules\\babel-loader\\lib\\index.js',
                    options:
                     { babelrc: false,
                       presets:
                        [ 'E:\\project\\presentation\\node_modules\\babel-preset-react-app\\index.js' ],
                       plugins:
                        [ 'react-hot-loader/babel',
                          'babel-plugin-transform-react-pug',
                          'babel-plugin-emotion',
                          [ '@babel/plugin-proposal-decorators', { legacy: true } ],
                          [ 'E:\\project\\presentation\\node_modules\\react-scripts\\node_modules\\babel-plugin-named-asset-import\\index.js',
                            { loaderMap: { svg: { ReactComponent: 'svgr/webpack![path]' } } } ] ],
                       cacheDirectory: true,
                       highlightCode: true } } ] },
             { test: /\.js$/,
               use:
                [ { loader:
                     'E:\\project\\presentation\\node_modules\\thread-loader\\dist\\cjs.js',
                    options: { poolTimeout: Infinity } },
                  { loader:
                     'E:\\project\\presentation\\node_modules\\react-scripts\\node_modules\\babel-loader\\lib\\index.js',
                    options:
                     { babelrc: false,
                       compact: false,
                       presets:
                        [ 'E:\\project\\presentation\\node_modules\\babel-preset-react-app\\dependencies.js' ],
                       cacheDirectory: true,
                       highlightCode: true } } ] },
             // ... garbage ...
```
Yet this is something...dubious:
```js
 { test: /\.js$/,
               use:
                [ { loader:
                     'E:\\project\\presentation\\node_modules\\thread-loader\\dist\\cjs.js',
                    options: { poolTimeout: Infinity } },
                  { loader:
                     'E:\\project\\presentation\\node_modules\\react-scripts\\node_modules\\babel-loader\\lib\\index.js',
                    options:
                     { babelrc: false,
                       compact: false,
                       presets:
                        [ 'E:\\project\\presentation\\node_modules\\babel-preset-react-app\\dependencies.js' ],
                       cacheDirectory: true,
                       highlightCode: true } } ] },
```
I'd try and delete this entry to see if it's okay after all

**Comment by aphillipo** [comment]
I'm seeing this also.

**Comment by abologna-r7** [comment]
do we have any news on this?

**Comment by ematipico** [comment]
Without a repo to reproduce the issue it might be difficult to triage the issue

**Comment by abologna-r7** [comment]
So I figure it out by not using .babelrc at all, and instead using babel.config.js :)

**Comment by taystack** [comment]
> So I figure it out by not using .babelrc at all, and instead using babel.config.js :)

Could you elaborate a bit more on this?

**Comment by kiwicopple** [comment]
Here is an example `babel.config.js` (you should delete .babelrc):

```
module.exports = function(api) {
  api.cache(true)
  return {
    presets: ['next/babel'],
    plugins: [
      ['@babel/plugin-proposal-decorators', { legacy: true }],
      ['@babel/plugin-proposal-class-properties', { loose: true }],
    ],
  }
}
```

**Comment by quantumpotato** [comment]
@kiwicopple  I tried that, still get the error. 
I'm trying to follow https://medium.com/react-native-training/react-native-with-mobx-getting-started-ba7e18d8ff44

**Comment by PlanetGit** [comment]
@kiwicopple Thank you very much. I fixed the error which is `SyntaxError:  ...extend-blocks.js: Support for the experimental syntax 'jsx' isn't currently ` issue. I tried to find the solution to fix that but I couldn't fix its error. According to your recommendation, I tried that deleted .babelrc file and then configured it in the `babel.config.js` file. I was tried to fix its error all day. It is working very well and very cool. I could save my time and effort. **I really appreciate you sharing your experience.**
